src/libmongoc/src/mongoc/mongoc-crypt.c

```diff
@@ -658,6 +675,179 @@ _state_machine_run (_state_machine_t *state_machine,
    return ret;
 }
 
+/* _parse_one_tls_opts parses one TLS document.
+ * @iter is an iterator at the start of a KMS provider key/value pair.
+ * @out_opt is written to on success, and left unmodified on error.
+ * Returns false and sets @error on error. */
+static bool
+_parse_one_tls_opts (bson_iter_t *iter,
+                     mongoc_ssl_opt_t *out_opt,
+                     bson_error_t *error)
+{
+   bool ok = false;
+   const char *kms_provider;
+   bson_t tls_opts_doc;
+   const uint8_t *data;
+   uint32_t len;
+   bson_string_t *errmsg;
+   bson_iter_t permitted_iter;
+
+   errmsg = bson_string_new (NULL);
+   kms_provider = bson_iter_key (iter);
+
+   if (!BSON_ITER_HOLDS_DOCUMENT (iter)) {
+      bson_set_error (error,
+                      MONGOC_ERROR_CLIENT_SIDE_ENCRYPTION,
+                      MONGOC_ERROR_CLIENT_INVALID_ENCRYPTION_ARG,
+                      "Expected TLS options for %s to be a document, got: %s",
+                      kms_provider,
+                      _mongoc_bson_type_to_str (bson_iter_type (iter)));
+      goto fail;
+   }
+
+   bson_iter_document (iter, &len, &data);
+   if (!bson_init_static (&tls_opts_doc, data, len) ||
+       !bson_iter_init (&permitted_iter, &tls_opts_doc)) {
+      bson_set_error (error,
+                      MONGOC_ERROR_CLIENT_SIDE_ENCRYPTION,
+                      MONGOC_ERROR_CLIENT_INVALID_ENCRYPTION_ARG,
+                      "Error iterating into TLS options document for %s",
+                      kms_provider);
+      goto fail;
+   }
+
+   while (bson_iter_next (&permitted_iter)) {
+      const char *key = bson_iter_key (&permitted_iter);
+
+      if (0 ==
+          bson_strcasecmp (key, MONGOC_URI_TLSCERTIFICATEKEYFILEPASSWORD)) {
+         continue;
+      } else if (0 == bson_strcasecmp (key, MONGOC_URI_TLSCERTIFICATEKEYFILE)) {
+         continue;
+      } else if (0 == bson_strcasecmp (key, MONGOC_URI_TLSCAFILE)) {
+         continue;
+      }
+
+      bson_set_error (error,
+                      MONGOC_ERROR_CLIENT_SIDE_ENCRYPTION,
+                      MONGOC_ERROR_CLIENT_INVALID_ENCRYPTION_ARG,
+                      "Error setting TLS options for %s: %s is prohibited",
```

Comment:
Is "prohibited" the right word to use here?

Reply:
Oops. "prohibited" matches the intent, but not the wording required in the specification:

> Drivers SHOULD raise an error if insecure TLS options are set. The error MUST contain the message "Insecure TLS options prohibited".

Will update.